integration-tests/src/tests/client/block_corruption.rs

```diff
@@ -99,3 +104,175 @@ fn change_shard_id_to_invalid() {
+            KeyType::ED25519,
+            "test0",
+        ));
+
+        if is_breaking_block_change(&correct_block, &corrupt_block) {
+            match env.clients[0].process_block_test(corrupt_block.clone().into(), Provenance::NONE)
+            {
+                Ok(_) => {
+                    return Err(anyhow::anyhow!(
+                        "Was able to process default block with {} bit switched.",
+                        corrupted_bit_idx
+                    ));
+                }
+                Err(e) => {
+                    if let Err(e) =
+                        env.clients[0].process_block_test(correct_block.into(), Provenance::NONE)
+                    {
+                        return Err(anyhow::anyhow!("Was unable to process default block after attempting to process default block with {} bit switched. {}",
+                        corrupted_bit_idx, e)
+                        );
+                    }
+                    return Ok(e.into());
+                }
+            }
+        } else {
+            return Ok(anyhow::anyhow!(NOT_BREAKING_CHANGE_MSG));
+        }
+    }
+    return Ok(anyhow::anyhow!(BLOCK_NOT_PARSED_MSG));
+}
+
+/// Produce a block. Flip a bit in it.
+/// Check that corrupt block cannot be processed. Check that correct block can be processed after.
+/// Do it for every bit in the block independently.
+///
+/// Checks are performed for all bits even if some of them fail.
+/// Results are accumulated in `errs` and `oks` vectors, that are printed at the end of the test.
+/// Test fails if
+/// - `errs` is not empty which means that for some bit either corrupt block was processed, or correct block was not.
+/// - `oks` only contain trivial errors which means that nothing was checked actually.
+///
+/// `oks` are printed to check the sanity of the test.
+/// This vector should include various validation errors that correspond to data changed with a bit flip.
+#[test]
+#[cfg_attr(not(feature = "expensive_tests"), ignore)]
+fn check_process_flipped_block_fails() {
+    init_test_logger();
+    let mut corrupted_bit_idx = 0;
+    // List of reasons `check_process_flipped_block_fails_on_bit` returned `Err`.
+    // Should be empty.
+    let mut errs = vec![];
+    // List of reasons `check_process_flipped_block_fails_on_bit` returned `Ok`.
+    // Should contain various validation errors.
+    let mut oks = vec![];
+    loop {
+        let res = check_process_flipped_block_fails_on_bit(corrupted_bit_idx);
+        if let Ok(res) = &res {
+            if res.to_string() == "End" {
+                // `corrupted_bit_idx` is out of bounds for correct block length. Should stop iteration.
+                break;
+            }
+        }
+        match res {
+            Err(e) => errs.push(e),
+            Ok(o) => oks.push(o),
+        }
+        corrupted_bit_idx += 1;
+    }
+    tracing::info!("All of the Errors:");
+    for err in &errs {
+        tracing::info!("{:?}", err);
+    }
```

Comment:
nit: When debugging it, it would be useful to know the index as well.

Reply:
corrupt_bit_index is in the message. Added better handling of inner check results to the follow-up issue.